src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +673,500 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+                // Override the older results of the partitions with the newly
+                // queried results.
+                for (size_t i = 0; i < requests_map.size(); i++) {
+                    auto r_i = requests_map[i];
+                    auto& r = results[r_i];
+                    total_size -= r.data_size_bytes();
+                    r = std::move(q_results.results[i]);
+                    total_size += r.data_size_bytes();
+
+                    _last_visible_indexes[r_i]
+                      = q_results.last_visible_indexes[i];
+                }
+            }
+
+            if (
+              total_size >= _ctx.min_bytes || q_results.has_error
+              || _as.abort_requested()) {
+                co_return worker_result{
+                  .read_results = std::move(results),
+                  .total_size = total_size,
+                };
+            }
+
+            std::optional<size_t> has_errored_request;
+            if (first_run) {
+                has_errored_request = register_waiters(
+                  std::ranges::iota_view{0ull, _ctx.requests.size()});
+            } else {
+                has_errored_request = register_waiters(requests_map);
+            }
+
+            if (has_errored_request) {
+                auto r_i = has_errored_request.value();
+                total_size -= results[r_i].data_size_bytes();
+                results[r_i] = read_result(
+                  error_code::not_leader_for_partition);
+                results[r_i].partition
+                  = _ctx.requests[r_i].ktp().get_partition();
+                co_return worker_result{
+                  .read_results = std::move(results),
+                  .total_size = total_size,
+                };
+            }
+
+            co_await _completed_waiter_count.wait();
+
+            if (_as.abort_requested()) {
+                co_return worker_result{
+                  .read_results = std::move(results),
+                  .total_size = total_size,
+                };
+            }
+
+            first_run = false;
+        }
+    }
+};
+
+/*
+ * A fetch exeutor that relies on notifications from raft to determine when to
+ * query partitions rather than querying all partitions at a set polling
+ * interval like the `parallel_fetch_plan_executor`.
+ */
+class nonpolling_fetch_plan_executor final : public fetch_plan_executor::impl {
+public:
+    nonpolling_fetch_plan_executor()
+      : _last_result_size(ss::smp::count, 0) {}
+
+    /**
+     * Executes the supplied `plan` until `octx.should_stop_fetch` returns true.
+     */
+    ss::future<> execute_plan(op_context& octx, fetch_plan plan) final {
+        start_worker_aborts(plan);
+
+        co_await log_and_ignore_exceptions(
+          do_execute_plan(octx, std::move(plan)));
+
+        // Send abort signal to workers and wait for all workers to end before
+        // returning.
+        co_await abort_workers();
+        co_await _workers_gate.close();
```

Comment:
Same exception safety thing here.

Reply:
If either of these throw an exception(which I don't believe can happen) then there is no way to recover and there is a potential for orphaned workers. It's possible to catch and log the exception here. However, there is a general try/catch a bit further up in the general Kafka request call stack that does the same. I can't remember if it logs the exception with an error or warn level though.

Reply:
ISTM that an orphaned worker is better than UAF UB. In any case the orphaned worker has already happened at the decision point we are talking about so the question is do you want a UAF on top of it?

Seems like a catch(...) { close; throw } is called for?

Reply:
Related thread: https://redpandadata.slack.com/archives/C044RD18NMV/p1706892629606879

Reply:
@ballard26 pointed out to me that the workers gate is entered in `do_execute_plan` which itself is wrapped to log + ignore exceptions and then `abort_workers` definitely can't throw because the underlying method called on each shard is `nothrow`, so I guess assuming the non-exception behavior to keep the code simple here is OK, so we should probably comment it.